drivers: uart: Fix incorrect len and offset in nrf UARTE async API.

Mierunski opened the pull request:
Due to longer than expected user callback handling, rx byte counting
got out of sync with real values. It leads to incorrect values
reported to user. This fix adds sync point at the end of buffer.

When using hardware rx counting this issue should not occur.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Fixes #17870

Mierunski committed:
drivers: uart: Fix incorrect len and offset in nrf UARTE async API.

Due to longer than expected user callback handling, rx byte counting
got out of sync with real values. It leads to incorrect values
reported to user. This fix adds sync point at the end of buffer.

When using hardware rx counting this issue should not occur.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski commented:
FYI @YuraCobain @heypiotr
I wasn't able to replicate issue with hardware counting enabled.

> However, as I already mentioned changing condition from not equal to less than in function:
> ```
>static void rx_timeout(struct k_timer timer)
>{
>...
> / Check if there is data that was not sent to user yet */
> if (data->async->rx_total_byte_cnt
> **>** data->async->rx_total_user_byte_cnt) { ...}
> ```
>
>works for me.

Changing this condition won't eliminate the cause of this bug. Condition has to stay this way in case when byte counter overflows.

YuraCobain commented:
Unfortunately enabling hardware counting does not resolve the issue.
My serial driver configuration is following:

 Serial Drivers


CONFIG_UART_NRFX=y
CONFIG_UART_0_NRF_UARTE=y
CONFIG_UART_0_ASYNC=y
CONFIG_UART_0_NRF_TX_BUFFER_SIZE=32
CONFIG_UART_0_NRF_HW_ASYNC=y
CONFIG_UART_0_NRF_HW_ASYNC_TIMER=1
CONFIG_NRFX_TIMER1=y
CONFIG_NRFX_TIMER=y
CONFIG_NRFX_PPI=y
CONFIG_UARTE_NRF_HW_ASYNC=y
CONFIG_NRF_UARTE_PERIPHERAL=y

But the bug still exists:
[00:00:00.662,872] <dbg> rfid_serial_transport.s_receiveBytes: rx ret:7 req:7 space:252 
[00:00:00.662,902] <dbg> rfid_serial_transport.s_receiveBytes: rx ret:3 req:3 space:255 
[00:00:00.665,496] <wrn> bt_hci_core: No ID address. App must call settings_load()
[00:00:00.666,625] <inf> bt_settings: Saving ID
[00:00:00.667,999] <inf> bt_settings: Saving ID
[00:00:00.680,969] <dbg> rfid_serial_transport.uart_dev_callback: rx_rdy space: 255 len:4294967170 off: 30
[00:00:00.680,969] <err> rfid_serial_transport: rb no space!

YuraCobain commented:
7b7c1a6089c441d839f61c7ea8008f5b8bd0a0ea does not fix the bug in my setup as well.

heypiotr commented:
I applied this patch locally, and disabled hardware counting (my previous workaround for this issue), and it fixed my problem (https://github.com/zephyrproject-rtos/zephyr/issues/17870#issuecomment-518666269). Thanks!

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -674,6 +674,13 @@ static void endrx_isr(struct device *dev)
+	data->async->rx_total_user_byte_cnt += rx_len;
+
+	if (!hw_rx_counting_enabled(data)) {
+		data->async->rx_cnt.cnt = data->async->rx_total_user_byte_cnt;
```
It would be good to add a comment here that this is done to prevent using an incorrect (too low) value of `rx_cnt.cnt` (which may occur due to latencies in handling of the `RXDRDY` interrupt) in the computation of the amount of new bytes to be passed to the user callback in `rx_timeout` (this is not that obvious). And I think it is also worth mentioning that because the whole buffer has been filled we can be sure here that the above value (`rx_total_user_byte_cnt`) is the current total number of received bytes.

anangl reviewed (approved):
The solution is good. I think it could be only described a bit better/at all.

ioannisg commented:
@nordic-krch could you take a look into this bug-fix?

Mierunski commented:
There is one more change I wan't to add, do not merge yet

ioannisg commented:
> 
> 
> There is one more change I wan't to add, do not merge yet

I would want, anyways, to get a +1 from @nordic-krch, as well

ioannisg commented:
This needs to go in for 2.0 - let's wait until @Mierunski can pick this up and finish it.

Mierunski committed:
drivers: nrf: Fix RX timeout and ENDRX ISR handling.

Ensure that two routines won't interfere with eachother.
In current situation there is possibility that ENDRX will be called
during rx_timeout routine or vice-versa which will result in wrong
offset and length passed to user.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski commented:
@YuraCobain Please check if current fix helps in your case

YuraCobain commented:
@Mierunski the fix resolved the issue. Thanks.
